src/components/tl/ucp/reduce_scatter/reduce_scatter_ring.c

```diff
@@ -0,0 +1,362 @@
+    size_t max_block_size, block_offset, frag_count, frag_offset, final_offset;
+    int    step, is_avg;
+    void * tmp[3];
+
+    final_offset = 0;
+    if (UCC_IS_INPLACE(*args)) {
+        sbuf = args->dst.info.buffer;
+        count /= size;
+        final_offset =
+            ucc_buffer_block_offset(count, size, UCC_TL_TEAM_RANK(team));
+    }
+
+    sendto   = ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, sendto);
+    recvfrom = ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, recvfrom);
+
+    max_block_size = ucc_buffer_block_count(count, size, 0) * dt_size;
+    tmp[0]         = task->reduce_scatter_ring.scratch;
+    tmp[1]         = PTR_OFFSET(tmp[0], max_block_size);
+    tmp[2]         = PTR_OFFSET(tmp[1], max_block_size);
+
+    if (UCC_INPROGRESS == ucc_tl_ucp_test_ring(task)) {
+        return task->super.super.status;
+    }
+    while (task->recv_posted > 0) {
+        void *reduce_target = tmp[1 + (task->recv_completed - 1) % 2];
+        step                = task->send_posted;
+        prevblock           = (rank - 1 - step + size) % size;
+        prevblock =
+            ucc_ep_map_eval(task->reduce_scatter_ring.inv_map, prevblock);
+        if (task->recv_completed == size - 1) {
+            ucc_assert(prevblock == UCC_TL_TEAM_RANK(team));
+            prevblock = UCC_TL_TEAM_RANK(team);
+        }
+        /* reduction */
+        ucc_assert(task->recv_posted == task->recv_completed);
+        ucc_assert(task->recv_posted < size);
+
+        ucc_ring_frag_count(task, count, prevblock, &frag_count);
+        ucc_ring_frag_block_offset(task, count, prevblock, &block_offset,
+                                   &frag_offset);
+        if (task->recv_completed == size - 1) {
+            reduce_target = PTR_OFFSET(args->dst.info.buffer,
+                                       (frag_offset + final_offset) * dt_size);
+        }
+        is_avg =
+            (args->op == UCC_OP_AVG) && (task->recv_completed == (size - 1));
+        if (UCC_OK !=
+            (status = ucc_tl_ucp_reduce_multi(
+                 tmp[0],
+                 PTR_OFFSET(sbuf, (block_offset + frag_offset) * dt_size),
+                 reduce_target, 1, frag_count, 0, dt, mem_type, task,
+                 is_avg))) {
+            tl_error(UCC_TASK_LIB(task), "failed to perform dt reduction");
+            task->super.super.status = status;
+            return status;
+        }
+        if (task->recv_completed == size - 1) {
+            task->recv_posted = task->recv_completed = 0;
+            break;
+        }
+        ucc_assert(task->send_posted - task->send_completed <= 1);
+        ucc_assert(task->send_posted < size);
+
+        UCPCHECK_GOTO(ucc_tl_ucp_send_nb(tmp[1 + (task->send_posted - 1) % 2],
+                                         frag_count * dt_size, mem_type, sendto,
+                                         team, task),
+                      task, out);
+
+        recv_data_from = (rank - 2 - step + size) % size;
```

Comment:
looks like this value is negative for rank 0 on last step
last step recv_data_from = (rank - 2 - (size - 1) + size) % size = (rank - 1) % size

Reply:
it never gets there. on last step the loop will break at line 132.